### PR TITLE
Add robot system interface and controller

### PR DESCRIPTION
**Pull request by ipa-cmh**
This pull request redesigns the ros2_control infrastructure, as a number of short comings have appeared during testing with real robots.

There is a need for a robot controller. The device controller is nice, but only interesting for very small setups. A robot system interface that automatically makes all drives of the robot ready to use is necessary.

### CODE REVIEW

File: canopen_ros2_controllers/include/canopen_ros2_controllers/cia402_robot_controller.hpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+   * @param time
+   * @param period
+   * @return controller_interface::return_type
+   */
+  CANOPEN_ROS2_CONTROLLERS__VISIBILITY_PUBLIC
+  controller_interface::return_type update(
+    const rclcpp::Time & time, const rclcpp::Duration & period) override;
+
+protected:
+  rclcpp::Service<std_srvs::srv::Trigger>::SharedPtr handle_recover_service_;
```
Comment (livanov93) [inline canopen_ros2_controllers/include/canopen_ros2_controllers/cia402_robot_controller.hpp:137]: Not used in the class implementation

File: canopen_ros2_controllers/src/cia402_robot_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        params_.operation_mode, joint_name.c_str());
+      // return controller_interface::CallbackReturn::ERROR;
+    }
+    command_interfaces_[jn + OPERATION_MODE].set_value(std::numeric_limits<double>::quiet_NaN());
+    command_interfaces_[jn + OPERATION_MODE_FEEDBACK].set_value(
+      std::numeric_limits<double>::quiet_NaN());
+    jn += ADD_OP;
+  }
+  RCLCPP_INFO(get_node()->get_logger(), "activate successful");
+  timer_->cancel();
```
Comment (livanov93) [inline canopen_ros2_controllers/src/cia402_robot_controller.cpp:172]: why are this commands only available in `on_activate`?

File: canopen_ros2_control/src/robot_system.cpp [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+  executor_.reset();
+  spin_thread_.reset();
+  motor_data_.clear();
+  cia402_drivers_.clear();
+}
+
+#include "pluginlib/class_list_macros.hpp"
+
+PLUGINLIB_EXPORT_CLASS(canopen_ros2_control::Cia402RobotSystem, hardware_interface::SystemInterface)
```
Comment (livanov93) [inline canopen_ros2_control/src/robot_system.cpp:308]: Most of this file is copied from [here](https://github.com/ros-industrial/ros2_canopen/blob/master/canopen_ros2_control/src/cia402_system.cpp), but there is no license mention nor the any link about the authors...

File: canopen_tests/launch/robot_control_setup.launch.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+from launch import LaunchDescription
```
Comment (livanov93) [inline canopen_tests/launch/robot_control_setup.launch.py:1]: Again copied from [here](https://github.com/ros- industrial/ros2_canopen/blob/master/canopen_ros2_control/launch/cia402_system.launch.py) and no license presented.

**Review by livanov93 — CHANGES_REQUESTED** [review]
Stuff implemented in the `Cia402RobotController` in `on_activate` and `on_deactivate` could have been implemented in the existing device controller. 
When initially implementing device controller, we mentioned that enabling multiple joints as parameters is great, which you did, but it could have be done within device controller and by changing it, and not reusing most of it by introducing new controller.
Also `Cia402RobotSystem` is reusing most of the stuff from already existing `Cia402System`. 
Most of the stuff added, should target changing existing files.

Best option would be to do switching between modes in `prepare_switch` and `perform_switch` while having separate command interfaces per each canopen mode.

**Comment by ipa-cmh** [comment]
@livanov93 sorry did not yet add **any license** as you can see in the files. Did not want to step on anyone's feet here and would not merge it as is.
We actually implemented this because we were testing our 6 axis robot with it and activation just gets nasty and kind of unstable with single device controllers. I also heard that some people are already using the control stack with their devices. As I could not foresee if we do breaking changes, I started a new/copy interface. In the end we can merge and maybe deprecate the others at some point or bring over the features to the other interfaces and kill the new one. 

**Robot System Interface**
Yes, we could add these features in CIA402System or CANopenSystem. However, the bigger goal is to move from an inheritance to a composition approach for the hardware interface (one interface that covers all device types). This is not yet done here and we only focussed on CIA402 devices, because we need them for tests. In the end the robot system interface should read the urdf and check which node_ids are indicated per joint. Then fetch the right driver type from bus.yml and use that for casting and handling.

**Robot Controller (probably obsolete)**:
We thought the robot controller should take care of bringing up the robots drives and settting the correct operation mode on activation. After discussion with Denis, I learned that this should rather be done automatically in hardware interface using perform_switch_mode and on_activate. So in the end the robot controller is obsolete and will be integrated into the system interface. Also the whole mode switching stuff in the other controllers should be obsolete.

**General observations reagrding controllers and driver nodes**:
* The driver nodes provide debugging functionality, diagnostics etc (services etc. for writing to the nodes manually). These should not be duplicated by a controller.
* The controllers should serve a specific purpose (i.e. forward_command_controller, ...)
* It is possible to use a controller as a kind of driver via the sdo/pdo interfaces that are in the CANopenSystem interface, but this should only ever be done for very simple devices. The cannonical way would be implementing a driver in the ros2_canopen framework.

**Pull request closed without merging**